### ISSUE TITLE
Non-functionally defined variable not in search annotation

### ISSUE THREAD

**Issue by erikvader** [erikvader/MiniZinc-linter#9]
# Explanation
If a search annotation is given should all non-functionally defined variables be included in it.

# Examples
`b` is functionally defined by `a`, so `a` should be included in a search annotation if one exists.
```minizinc
var 1..2: a;
var int: b;
array[1..2] of int: A = [42, 43];
constraint b = A[a];
```

**Comment by erikvader** [comment]
Should only non-functionally variables be used in search hints? Can functionally defined ones also be included? Why was this a problem again?

**Comment by matsc-at-sics-se** [comment]
Why on earth should all non-functionally defined variables have to be included? Many solvers have excellent default handling of variables not in the search annotation. Also, let-variables can be non-functionally defined, for example in some rules of the solver libraries. Such variables can't be included by construction.

**Comment by GustavBjordal** [comment]
@matsc-at-sics-se This is originating from one of the [MiniZinc Challenge rules](https://www.minizinc.org/challenge2020/rules2020.html) namely that "Each solve item must be annotated with a search strategy, such that fixing all the variables appearing in the search strategy would allow a value propagation solver to check a solution."
A rough approximation of this is to say that each variable that is not functionally defined must appear in the search strategy. 
But I guess this is linter rule of low priority.

**Comment by erikvader** [comment]
This came from my notes from the startup meeting, and I didn't really understand it myself. I will maybe not start with this one.

**Comment by Pierre-Flener** [comment]
Yes, this is only for the --challenge option, but highest priority!
